File: src/pytypes/onefuzztypes/webhooks.py
#!/usr/bin/env python
#
# Copyright (c) Microsoft Corporation.
# Licensed under the MIT License.

from datetime import datetime
from enum import Enum
from typing import List, Optional
from uuid import UUID, uuid4

from pydantic import AnyHttpUrl, BaseModel, Field

from .enums import WebhookMessageState
from .events import Event, EventMessage, EventType


class WebhookMessageFormat(Enum):
    onefuzz = "onefuzz"
    event_grid = "event_grid"


class WebhookMessage(EventMessage):
    webhook_id: UUID


class WebhookMessageEventGrid(BaseModel):
    dataVersion: str
    subject: str
    eventType: EventType
    eventTime: datetime
    id: UUID
    data: Event


class WebhookMessageLog(WebhookMessage):
    state: WebhookMessageState = Field(default=WebhookMessageState.queued)
    try_count: int = Field(default=0)


class Webhook(BaseModel):
    webhook_id: UUID = Field(default_factory=uuid4)
    name: str
    url: Optional[AnyHttpUrl]
    event_types: List[EventType]
    secret_token: Optional[str]
    message_format: Optional[WebhookMessageFormat]

File: src/ApiService/ApiService/OneFuzzTypes/Webhooks.cs
using System.Text.Json.Serialization;
using Microsoft.OneFuzz.Service.OneFuzzLib.Orm;

namespace Microsoft.OneFuzz.Service;


public enum WebhookMessageFormat {
    Onefuzz,
    EventGrid
}

public record WebhookMessageQueueObj(
        Guid WebhookId,
        Guid EventId
        );

public record WebhookMessage(Guid EventId,
    EventType EventType,
    BaseEvent Event,
    Guid InstanceId,
    String InstanceName,
    Guid WebhookId,
    DateTime CreatedAt,
    Uri SasUrl,
    DateOnly? ExpiresOn = null) : DownloadableEventMessage(EventId, EventType, Event, InstanceId, InstanceName, CreatedAt, SasUrl, ExpiresOn);


// the schema for the event grid message is defined here:
// https://learn.microsoft.com/en-us/azure/event-grid/event-schema#event-schema
public record WebhookMessageEventGrid(
    [property: JsonPropertyName("dataVersion")] string DataVersion,
    string Subject,
    [property: JsonPropertyName("EventType")] EventType EventType,
    [property: JsonPropertyName("eventTime")] DateTimeOffset EventTime,
    Guid Id,
    WebhookMessage Data
    );

public record WebhookMessageLog(
    [RowKey] Guid EventId,
    EventType EventType,
    [property: TypeDiscrimnator("EventType", typeof(EventTypeProvider))]
    [property: JsonConverter(typeof(BaseEventConverter))]
    BaseEvent Event,
    Guid InstanceId,
    String InstanceName,
    [PartitionKey] Guid WebhookId,
    long TryCount,
    WebhookMessageState State = WebhookMessageState.Queued
    ) : EntityBase();

public record Webhook(
    [PartitionKey] Guid WebhookId,
    [RowKey] string Name,
    Uri? Url,
    List<EventType> EventTypes,
    string? SecretToken, // SecretString??
    WebhookMessageFormat? MessageFormat
    ) : EntityBase();
